### DISCUSSION TITLE
How symlinked node_modules structure helps to avoid circular symlinks?

### DISCUSSION THREAD

**Discussion by Peccansy** [pnpm/pnpm#3554] [Q&A] [answered]
I want to clarify for myself this statement from the FAQ:
> Although pnpm uses linking to put dependencies into node_modules folders, circular symlinks are avoided because parent packages are placed into the same node_modules folder in which their dependencies are. So foo's dependencies are not in foo/node_modules, but foo is in node_modules together with its own dependencies." 

Please, can somebody give an example of a problem? I think that I don't understand in which case a cycle might appear when we work with different structure :(

**Comment by zkochan** [comment, marked as answer]
A simple example would be: foo has bar in the dependencies and bar has foo in the dependencies. foo -> bar -> foo -> bar...

**Reply by zkochan** [reply]
if pnpm would write the node_modules of foo to `foo/node_modules`, we would get a circular symlink

```
foo/node_modules/bar -> ../../bar
bar/node_modules/foo -> ../../foo
```

but pnpm writes node_modules one level up

```
foo@1.0.0/node_modules/bar -> ../../bar@1.0.0/node_modules/bar
bar@1.0.0/node_modules/foo -> ../../foo@1.0.0/node_modules/foo
```

**Reply by zkochan** [reply]
related pages:
* https://pnpm.io/symlinked-node-modules-structure
* https://pnpm.io/blog/2020/05/27/flat-node-modules-is-not-the-only-way
